File: implementations/javascript/src/parser/visitor.ts
import { LSFNode, ParseResult } from './types';
import { DOMNavigator as DOMNavigatorClass } from './dom-navigator';
import { NODE_TYPE } from './dom-builder'; // Need NODE_TYPE constants

// Simple StringBuilder for efficient string concatenation
class StringBuilder {
    private parts: string[] = [];
    private length: number = 0;
    // TODO: Consider pre-allocation based on estimated size

    append(str: string): void {
        this.parts.push(str);
        this.length += str.length;
    }

    toString(): string {
        return this.parts.join('');
    }

    getLength(): number {
        return this.length;
    }
}

// Visitor Interface (as per plan)
export interface Visitor {
  visitObject(nodeIndex: number): void;
  visitField(nodeIndex: number): void;
  visitValue(nodeIndex: number): void;
  getResult(): any; // Return type depends on visitor (string for JSON)
}

// LSF -> JSON Visitor Implementation
export class LSFToJSONVisitor implements Visitor {
  private result: StringBuilder;
  private navigator: DOMNavigatorClass;
  private parseResult: ParseResult;

  constructor(parseResult: ParseResult) {
    this.parseResult = parseResult;
    this.navigator = parseResult.navigator as DOMNavigatorClass;
    // TODO: Estimate output size based on input buffer size or node count?
    this.result = new StringBuilder();
  }

  // Public entry point to start traversal from the root
  buildJSON(): string {
      // Get all root indices to handle multiple objects
      const rootIndices = this.navigator.getRootIndices();
      
      if (rootIndices.length === 0) {
          // No root objects found
          return "{}";
      }
      
      if (rootIndices.length === 1) {
          // Single root object - return as single JSON object
          this.visit(rootIndices[0]);
          return this.getResult();
      }
      
      // Multiple root objects - return as JSON array
      this.result.append('[');
      for (let i = 0; i < rootIndices.length; i++) {
          if (i > 0) {
              this.result.append(',');
          }
          this.visit(rootIndices[i]);
      }
      this.result.append(']');
      return this.getResult();
  }
  
  // Generic visit method to dispatch based on node type
  private visit(nodeIndex: number): void {
      const nodeType = this.navigator.getType(nodeIndex);
      switch (nodeType) {
          case NODE_TYPE.OBJECT:
              this.visitObject(nodeIndex);
              break;
          case NODE_TYPE.FIELD:
              this.visitField(nodeIndex);
              break;
          case NODE_TYPE.VALUE:
              this.visitValue(nodeIndex);
              break;
          default:
              console.error(`Unknown node type encountered: ${nodeType} at index ${nodeIndex}`);
              // Optionally throw an error
              break;
      }
  }

  visitObject(nodeIndex: number): void {
    // TODO: Implement Object -> { ... } conversion
    this.result.append('{');
    const children = this.navigator.getChildren(nodeIndex);
    let firstField = true;
    for (const childIndex of children) {
        // Object children should be Fields
        if (this.navigator.getType(childIndex) === NODE_TYPE.FIELD) {
            if (!firstField) {
                this.result.append(',');
            }
            this.visitField(childIndex);
            firstField = false;
        } else {
             console.error(`Object node ${nodeIndex} has non-field child ${childIndex}`);
        }
    }
    this.result.append('}');
  }

  visitField(nodeIndex: number): void {
    // TODO: Implement Field -> "name": value or "name": [value1, value2]
    const fieldName = this.navigator.getName(nodeIndex);
    // JSON requires keys to be strings, escape fieldName
    this.result.append(JSON.stringify(fieldName || '__implicit_field__')); // Use placeholder for implicit
    this.result.append(':');

    const children = this.navigator.getChildren(nodeIndex);
    if (children.length === 0) {
        this.result.append('null'); // Field with no values -> null?
    } else if (children.length === 1) {
        // Single value
        this.visitValue(children[0]);
    } else {
        // Implicit array
        this.result.append('[');
        let firstValue = true;
        for (const childIndex of children) {
             if (this.navigator.getType(childIndex) === NODE_TYPE.VALUE) {
                if (!firstValue) {
                    this.result.append(',');
                }
                this.visitValue(childIndex);
                firstValue = false;
            } else {
                 console.error(`Field node ${nodeIndex} has non-value child ${childIndex}`);
            }
        }
        this.result.append(']');
    }
  }

  visitValue(nodeIndex: number): void {
    // TODO: Implement Value -> JSON primitive based on type hint
    const typeHint = this.navigator.getTypeHint(nodeIndex);
    const rawValue = this.navigator.getValue(nodeIndex); // Get decoded string value

    switch (String.fromCharCode(typeHint)) {
      case 'n': // number (integer)
        // Try parsing as int, fallback to float? Or just use parseFloat?
        const intVal = parseInt(rawValue, 10);
        this.result.append(isNaN(intVal) ? 'null' : String(intVal)); 
        break;
      case 'f': // float
        const floatVal = parseFloat(rawValue);
        this.result.append(isNaN(floatVal) ? 'null' : String(floatVal));
        break;
      case 'b': // boolean
        // Define LSF boolean representation ('true'/'false', '1'/'0'?)
        // Assuming 'true'/'false' case-insensitive for now
        const boolVal = rawValue.toLowerCase() === 'true';
        this.result.append(String(boolVal));
        break;
      case 'd': // date -> ISO string?
        // TODO: Define LSF date format (e.g., epoch ms, ISO string)
        // Assuming ISO 8601 string for now
        // Add validation? Fallback?
        this.result.append(JSON.stringify(rawValue)); // Treat as string for now
        break;
      case 'z': // null
        this.result.append('null');
        break;
      case 's': // string
      default: // Default to string
        this.result.append(JSON.stringify(rawValue)); // Includes escaping
        break;
    }
  }

  getResult(): string {
    return this.result.toString();
  }
} 

File: implementations/python/lsf/parser/visitor.py
"""Visitor pattern implementation for LSF to JSON conversion."""
import json
from typing import List, Optional, Any
from .types import TokenType, ValueHint
from .dom_navigator import DOMNavigator


class LSFToJSONVisitor:
    """Converts LSF DOM to JSON string."""
    
    def __init__(self, navigator: DOMNavigator):
        """Initialize visitor with navigator."""
        self.navigator = navigator
        self.result: List[str] = []
    
    def to_json_string(self) -> str:
        """Convert LSF DOM to JSON string."""
        root_indices = self.navigator.get_root_indices()
        
        if not root_indices:
            return '{}'
        
        if len(root_indices) == 1:
            # Single root object
            self.visit_object(root_indices[0])
        else:
            # Multiple root objects - return as array
            self.result.append('[')
            for i, root_index in enumerate(root_indices):
                if i > 0:
                    self.result.append(',')
                self.visit_object(root_index)
            self.result.append(']')
        
        return ''.join(self.result)
    
    def visit_object(self, node_index: int) -> None:
        """Visit an object node."""
        self.result.append('{')
        
        children = self.navigator.get_children_indices(node_index)
        first_field = True
        
        for child_index in children:
            child = self.navigator.get_node(child_index)
            if child and child.type == TokenType.FIELD:
                if not first_field:
                    self.result.append(',')
                self.visit_field(child_index)
                first_field = False
        
        self.result.append('}')
    
    def visit_field(self, node_index: int) -> None:
        """Visit a field node."""
        # Get field name
        field_name = self.navigator.get_node_data_as_string(node_index)
        if not field_name:
            field_name = '__implicit_field__'
        
        # Add field name
        self.result.append(json.dumps(field_name))
        self.result.append(':')
        
        # Get children (values)
        children = self.navigator.get_children_indices(node_index)
        
        if not children:
            self.result.append('null')
        elif len(children) == 1:
            # Single value
            self.visit_value(children[0])
        else:
            # Multiple values - array
            self.result.append('[')
            for i, child_index in enumerate(children):
                if i > 0:
                    self.result.append(',')
                self.visit_value(child_index)
            self.result.append(']')
    
    def visit_value(self, node_index: int) -> None:
        """Visit a value node."""
        node = self.navigator.get_node(node_index)
        if not node:
            self.result.append('null')
            return
        
        raw_value = self.navigator.get_node_data_as_string(node_index)
        
        # Apply type hint
        if node.type_hint == ValueHint.NUMBER:
            try:
                self.result.append(str(int(raw_value)))
            except ValueError:
                self.result.append('null')
        elif node.type_hint == ValueHint.FLOAT:
            try:
                self.result.append(str(float(raw_value)))
            except ValueError:
                self.result.append('null')
        elif node.type_hint == ValueHint.BOOLEAN:
            if raw_value.lower() == 'true':
                self.result.append('true')
            elif raw_value.lower() == 'false':
                self.result.append('false')
            else:
                self.result.append('null')
        elif node.type_hint == ValueHint.NULL:
            self.result.append('null')
        else:
            # String or no type hint
            self.result.append(json.dumps(raw_value))